### PR TITLE
`process.GetInfoForPid` collects the total thread count

### PR DESCRIPTION
**Pull request by AndersonQ**
## What does this PR do?
Collect the total thread count, num_threads, as part of the information collected by `process.GetInfoForPid`

* On linux:
  -  it extends the information parsed from /proc/[PID]/stat to get the total thread count
  -  unit tests added, they mock /proc/[PID]/stat
* On darwin:
  - it uses the syscall proc_pidinfo
  - unit test only validates GetInfoForPid does not return an error and the total thread count ins't zero
* On windows:
  - it uses the Process Snapshotting APIs: https://learn.microsoft.com/en-us/previous-versions/windows/desktop/proc_snap/overview-of-process-snapshotting
  - unit test only validates GetInfoForPid does not return an error and the total thread count ins't
* Add a unit test using a C++ program which runs exactly 42 threads so we can assert the data read by process.GetInfoForPid On windows the and newer the process start with more threads than the program requests to parallel load needed tabels/dll or some windows specific things. Therefore the test on windows accepts 42 or 45 for the number of threads allocated to the process. See the following for details https://stackoverflow.com/questions/42789199/why-there-are-three-unexpected-worker-threads-when-a-win32-console-application-s/42789684#42789684

## Why is it important?

To enable metricbeat to collect the total thread count per process as requested by https://github.com/elastic/beats/issues/34461.

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.md`

## Related issues

- Relates to https://github.com/elastic/beats/issues/34461

### CODE REVIEW

File: metric/system/process/process_linux_common.go [hunk 1]
```diff
@@ -121,64 +120,81 @@ func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter fu
 		return state, fmt.Errorf("error getting metadata for pid %d: %w", pid, err)
 	}
 
-	//username
+	// username
```
Comment (rdner) [inline metric/system/process/process_linux_common.go:123]: This comment does not appear to be useful.
```suggestion
```

File: metric/system/process/process_linux_common.go [hunk 2]
```diff
@@ -121,64 +120,81 @@ func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter fu
...
 	if err != nil {
 		if os.IsNotExist(err) {
 			return ProcState{}, syscall.ESRCH
 		}
 		return ProcState{}, fmt.Errorf("error reading procdir %s: %w", path, err)
+	}
 
+	state, err := parseProcStat(data)
+	state.Pid = opt.IntWith(pid)
+	if err != nil {
```
Comment (rdner) [inline metric/system/process/process_linux_common.go:146]: Why do we check the error after we assign to the `state`?

File: metric/system/process/process_linux_common.go [hunk 3]
```diff
@@ -121,64 +120,81 @@ func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter fu
...
-		return state, fmt.Errorf("failed to parse stat fields for pid %d from '%v': %w", pid, string(data), err)
+		return state, fmt.Errorf("failed to parse stat fields from '%s': %w",
+			string(data), err)
 	}
+
 	state.State = getProcState(procState[0])
 	state.Ppid = opt.IntWith(ppid)
 	state.Pgid = opt.IntWith(pgid)
-	state.Pid = opt.IntWith(pid)
+	state.NumThreads = opt.IntWith(numThreads)
```
Comment (rdner) [inline metric/system/process/process_linux_common.go:197]: Why did you move `state.Pid = opt.IntWith(pid)` to line 145?

File: metric/system/process/process_test.go [hunk 4]
```diff
@@ -122,7 +178,7 @@ func TestNetworkFilter(t *testing.T) {
 }
 
 func TestFilter(t *testing.T) {
-	//The logic itself is os-independent, so we'll only test this on the platform least likly to have CI issues
+	// The logic itself is os-independent, so we'll only test this on the platform least likly to have CI issues
```
Comment (rdner) [inline metric/system/process/process_test.go:181]:
```suggestion
	// The logic itself is os-independent, so we'll only test this on the platform least likely to have CI issues
```

File: metric/system/process/process_windows.go [hunk 5]
```diff
@@ -50,25 +52,65 @@ func (procStats *Stats) FetchPids() (ProcsMap, []ProcState, error) {
...
+	}
+
+	if numThreads, err := FetchNumThreads(pid); err != nil {
+		errs = append(errs, fmt.Errorf("error fetching num threads: %w", err))
+	} else {
+		state.NumThreads = opt.IntWith(numThreads)
+	}
+
+	if err := errors.Join(errs...); err != nil {
+		return state, fmt.Errorf("could not get all informations for PID %d: %w",
```
Comment (rdner) [inline metric/system/process/process_windows.go:81]:
```suggestion
		return state, fmt.Errorf("could not get all information for PID %d: %w",
```

File: metric/system/process/process_linux_common.go [hunk 6]
```diff
@@ -121,64 +120,81 @@ func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter fu
...
+func parseProcStat(data []byte) (ProcState, error) {
+	const minFields = 36
 	state := ProcState{}
 
 	// Extract the comm value with is surrounded by parentheses.
 	lIdx := bytes.Index(data, []byte("("))
 	rIdx := bytes.LastIndex(data, []byte(")"))
 	if lIdx < 0 || rIdx < 0 || lIdx >= rIdx || rIdx+2 >= len(data) {
-		return state, fmt.Errorf("failed to extract comm for pid %d from '%v': %w", pid, string(data), err)
+		return state, fmt.Errorf("failed to extract comm from '%v'",
```
Comment (rdner) [inline metric/system/process/process_linux_common.go:161]:
```suggestion
		return state, fmt.Errorf("failed to extract 'comm' field from '%v'",
```

File: metric/system/process/process_linux_common.go [hunk 7]
```diff
@@ -121,64 +120,81 @@ func FillPidMetrics(hostfs resolve.Resolver, pid int, state ProcState, filter fu
...
 		if os.IsNotExist(err) {
 			return ProcState{}, syscall.ESRCH
 		}
 		return ProcState{}, fmt.Errorf("error reading procdir %s: %w", path, err)
+	}
 
+	state, err := parseProcStat(data)
+	state.Pid = opt.IntWith(pid)
+	if err != nil {
+		return state, fmt.Errorf("failed to parse infos for pid %d': %w", pid, err)
```
Comment (rdner) [inline metric/system/process/process_linux_common.go:147]:
```suggestion
		return state, fmt.Errorf("failed to parse information for pid %d': %w", pid, err)
```

File: metric/system/process/process_windows_test.go [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+
+	"github.com/elastic/elastic-agent-system-metrics/metric/system/resolve"
+)
+
+func TestGetInfoForPid_numThreads(t *testing.T) {
+	const want = 42
+	// On windows the programs always start with more threads than the code
```
Comment (rdner) [inline metric/system/process/process_windows_test.go:32]:
```suggestion
	// On windows programs always start with more threads than the code
```

File: metric/system/process/process_windows_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	// is launching. It seems to be due to it loading dll/tables in parallel
+	// or de default thread pool. See the following for more info:
+	// https://stackoverflow.com/questions/42789199/why-there-are-three-unexpected-worker-threads-when-a-win32-console-application-s/42789684#42789684
+	//
+	// Use the Process Explorer https://learn.microsoft.com/en-us/sysinternals/downloads/process-explorer
+	// to check the number of threads allocated to a process. To see the number
+	// of threads in use you'd need to add the Threads column yourself.
```
Comment (rdner) [inline metric/system/process/process_windows_test.go:39]:
```suggestion
	// of threads in use, you'd need to add the Threads column yourself.
```

File: metric/system/process/process.go [hunk 10]
```diff
@@ -343,7 +341,7 @@ func (procStats *Stats) includeTopProcesses(processes []ProcState) []ProcState {
 
 // isWhitelistedEnvVar returns true if the given variable name is a match for
 // the whitelist. If the whitelist is empty it returns false.
-func (procStats Stats) isWhitelistedEnvVar(varName string) bool {
+func (procStats *Stats) isWhitelistedEnvVar(varName string) bool {
```
[the remark this answers is not in the document]
Comment (AndersonQ) [reply metric/system/process/process.go:344]: I just made it consistent with the other receivers

File: metric/system/process/process_linux_common.go (hunk 2 above)
Comment (AndersonQ) [reply metric/system/process/process_linux_common.go:145]: it needs to be done regardless of the error. `parseProcStat` does not do that, but it returns the most data it can get, so the best moment to set it is after `parseProcStat` returns, that way I can add the PID to the `state` it returned.

File: metric/system/process/process_linux_common.go (hunk 3 above)
Comment (AndersonQ) [reply metric/system/process/process_linux_common.go:196]: not sure where is the best place to set it. bit I can put it back

File: metric/system/process/process.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (rdner) [reply metric/system/process/process.go:344]: up to you, for me it's not about consistency, it's about semantics, if I have a pointer receiver in a function that means this function changes the state. Otherwise, you indicate that the function is read-only.

**Review by rdner — APPROVED** [review]
LGTM, I'll leave decision on the pointer receiver to you, @AndersonQ

**Review by fearful-symmetry — APPROVED** [review]
LGTM
